src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs

```diff
@@ -0,0 +1,132 @@
+using System;
+using Dapper;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using OrchardCore.ContentManagement.Metadata;
+using OrchardCore.Data;
+using OrchardCore.Data.Migration;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Scope;
+using OrchardCore.Search.Lucene.Model;
+using YesSql;
+
+namespace OrchardCore.Search.Lucene
+{
+    public class Migrations : DataMigration
```

Comment:
I'm not sure about this one because we could have other Lucene migrations in the future. Of course, this is what we are doing in this migration because there is only one but I would prefer that this Migrations class be considered as the Lucene module one. 

What would be the advantage if I rename this file?

Reply:
Having multiple migrations in the future is exactly why I suggest this name: "Migrations" is just general, if we want to add anything in the future then we'll either have a) a `Migrations` class and `SomethingElseMigrations`, which is weird, or b) we'd perpetually extend this one class.

Reply:
We generally have a Migrations class in every module to initialize them. The only advantage I see with this is that this migration file could be dropped eventually. But we generally keep these migrations and all its history into a Migrations class that is global to its module.

Reply:
Those migrations are not about migrating from an old configuration schema to a new one though (as in, a backward compatibility-fixing migration), but creating content definitions and tables/indexes.

Reply:
Okay I see, as I understand would be to do specific things normally only required for an existing installation, so not tied to what normally should be done on a new installation, and in the future we may need such a migration Create.

So I understand about the name but because it is the only one migration I think that's Okay, and if in the future we need something else we will just have to add a migration step (and also update the Create() to do it and return the last version), this without having to create a new migration.

Normally this is these specific things that would need to be done in a migration step e.g. `UpdateFrom1()` and then update the `Create()` method, not to do it in that case but just to return 2, so that it is only executed on an existing installation. But we can't do that if the module/feature has no existing migration (here because it is a new module).

The only "problem" I see would be that this new Migration.Create() will be always executed on any new installation, but if the things to replace don't exist, I assume that the sql UPDATES don't fail, if so for me that's Okay.

Maybe in a ModularTenantEvent to make the difference (I think we already did this), but then executed on each shell build so not better. Maybe check if the things to replace exist, hmm but anyway would need a sql query.

**So in summary LGTM ;) And maybe intended to be removed in a future release.**

Oh yes I remember, maybe we did such things in `IFeatureEventHandler.InstalledAsync()` or `EnabledAsync()`.

---

Also I saw that you use for example `tablePrefix += '_'`, now the prefix separator is configurable through `YesSqlOptions`, also the `ITableNameConvention`, but the last one may be null in `YesSqlOptions` and should be resolved from the DI, fixed in  #12430 where you can't resolve it anymore but is always existing and in sync in `YesSqlOptions`.

That said here you resolve the `ISesion` and then have access to `session.Store.Configuration` wich configuration already has what we need e.g. `.TableNameConvention` on which we can call `.GetDocumentTable(collection)` which is already aware of the table prefix and separator.

Reply:
Okay I see, as I understand would be to do specific things normally only required for an existing installation, so not tied to what normally should be done on a new installation, and in the future we may need such a migration Create.

So I understand about the name but because it is the only one migration I think that's Okay, and if in the future we need something else we will just have to add a migration step (and also update the Create() to do it and return the last version), this without having to create a new migration.

Normally this is these specific things that would need to be done in a migration step e.g. `UpdateFrom1()` and then update the `Create()` method, not to do it in that case but just to return 2, so that it is only executed on an existing installation. But we can't do that if the module/feature has no existing migration (here because it is a new module).

The only "problem" I see would be that this new Migration.Create() will be always executed on any new installation, but if the things to replace don't exist, I assume that the sql UPDATES don't fail, if so for me that's Okay.

Maybe in a ModularTenantEvent to make the difference (I think we already did this), but then executed on each shell build so not better. Maybe check if the things to replace exist, hmm but anyway would need a sql query.

**So in summary LGTM ;) And maybe intended to be removed in a future release.**

Oh yes I remember, maybe we did such things in `IFeatureEventHandler.InstalledAsync()` or `EnabledAsync()`.

---

Also I saw that you use for example `tablePrefix += '_'`, now the prefix separator is configurable through `YesSqlOptions`, also the `ITableNameConvention`, but the last one may be null in `YesSqlOptions` and should be resolved from the DI, fixed in  #12430 where you can't resolve it anymore but is always existing and in sync in `YesSqlOptions`.

That said here you resolve the `ISesion` and then have access to `session.Store.Configuration` wich configuration already has what we need e.g. `.TableNameConvention` on which we can call `.GetDocumentTable(collection)` which is already aware of the table prefix and separator.

Reply:
Then it needs to be changed also in IndexingTaskManager.cs file.